### PR TITLE
[SPARK-30793][SQL] Fix truncations of timestamps before the epoch to minutes and seconds

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to replace `%` by `Math.floorMod` in `DateTimeUtils.truncTimestamp` for the `SECOND` and `MINUTE` levels.

### Why are the changes needed?
This fixes the issue of incorrect truncation of timestamps before the epoch `1970-01-01T00:00:00.000000Z` to the `SECOND` and `MINUTE` levels. For example, timestamps after the epoch are truncated by cutting off the rest part of the timestamp:
```sql
spark-sql> select date_trunc('SECOND', '2020-02-11 00:01:02.123');
2020-02-11 00:01:02
```
but seconds in the truncated timestamp before the epoch are increased by 1:
```sql
spark-sql> select date_trunc('SECOND', '1960-02-11 00:01:02.123');
1960-02-11 00:01:03
```

### Does this PR introduce any user-facing change?
Yes. After the changes, the example above outputs correct result:
```sql
spark-sql> select date_trunc('SECOND', '1960-02-11 00:01:02.123');
1960-02-11 00:01:02
```

### How was this patch tested?
Added new tests to `DateFunctionsSuite`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@srowen @maropu @kiszk May I ask you to review this PR.

**Comment by kiszk** [comment]
This change looks good. In addition, I think that this function has the similar problems in `TRUNC_TO_HOUR` and `TRUNC_TO_DAY`, too.

How about the results with `val df = Seq("1961-04-12 03:01:02.345")...`?

**Comment by MaxGekk** [comment]
@kiszk Right, and truncations to hours and days have one more issue of getting time zone offsets. This should be fixed by another PR: https://github.com/apache/spark/pull/27512

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 1]
```diff
@@ -828,4 +828,17 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
       TimeZone.setDefault(defaultTz)
     }
   }
+
+  test("truncate timestamps before the epoch to seconds and minutes") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala]: Can you add the jira number in the prefix?

**Comment by maropu** [comment]
The code itself looks fine. btw, is this not a correctness issue? (trivial though)

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala]: Added JIRA id to the test title.

**Comment by MaxGekk** [comment]
> The code itself looks fine. btw, is this not a correctness issue? (trivial though)

I do believe it is a correctness issue. Rounding 2.123 -> 3 doesn't make any sense. This is not arithmetic rounding, and not just cutting off the rest. And this contradicts to truncations to other levels.

**Review by kiszk — APPROVED** [review]
Pending Jenkins result

**Comment by maropu** [comment]
cc: @dongjoon-hyun

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon Please, take a look at the PR.

**Comment by cloud-fan** [comment]
I've merged https://github.com/apache/spark/pull/27596 , let's see if tests still pass here.

**Comment by cloud-fan** [comment]
retest it please

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by cloud-fan** [comment]
to confirm, this bug doesn't exist in 2.4 as we switch the calendar in 3.0 only, right @MaxGekk ?

**Comment by MaxGekk** [comment]
@cloud-fan 2.4 **has** the same issue:
```sql
$ spark-sql -S
20/02/17 08:12:28 INFO SparkContext: Running Spark version 2.4.5
...
spark-sql> select date_trunc('SECOND', '1960-02-11 00:01:02.123');
1960-02-11 00:01:03
```

**Comment by cloud-fan** [comment]
cool, thanks!

**Comment by MaxGekk** [comment]
Here is the PR for 2.4: https://github.com/apache/spark/pull/27611

**Pull request closed without merging**